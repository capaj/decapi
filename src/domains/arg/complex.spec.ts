import { GraphQLNonNull, getNamedType, GraphQLString } from 'graphql'
import {
  InputField,
  ObjectType,
  Field,
  compileObjectType,
  InputObjectType,
  compileInputObjectType,
  Arg
} from '../..'

describe('Complex arguments', () => {
  it('should not allow complex argument type not decorated with @InputObjectType', async () => {
    class InputOne {
      @InputField()
      bar: string
    }

    @ObjectType()
    class Foo {
      @Field()
      bar(input: InputOne): string {
        return 'ok'
      }
    }
    expect(() =>
      compileObjectType(Foo).getFields()
    ).toThrowErrorMatchingSnapshot()
  })

  it('should not allow complex argument type decorated with @ObjectType', async () => {
    @ObjectType()
    class InputTwo {
      @Field()
      bar: string
    }

    @ObjectType()
    class Foo {
      @Field()
      bar(input: InputTwo): string {
        return 'ok'
      }
    }

    expect(() =>
      compileObjectType(Foo).getFields()
    ).toThrowErrorMatchingSnapshot()
  })
  it('Supports complex input types', async () => {
    @InputObjectType()
    class Input {
      @InputField()
      bar: string
    }

    @ObjectType()
    class Foo {
      @Field()
      nonNullable(input: Input): string {
        return 'ok'
      }
      @Field()
      nullable(input: Input | null): string {
        return 'ok'
      }
    }
    const { nonNullable, nullable } = compileObjectType(Foo).getFields()
    expect(nonNullable.args[0].type).toEqual(
      new GraphQLNonNull(compileInputObjectType(Input))
    )
    expect(nullable.args[0].type).toEqual(compileInputObjectType(Input))
  })

  it('Supports scalar list argument type', () => {
    @ObjectType()
    class Foo {
      @Field()
      bar(@Arg({ type: [String] }) input: string[]): string {
        return 'ok'
      }
    }
    const { bar } = compileObjectType(Foo).getFields()
    const argType = bar.args[0].type
    expect(argType.toString()).toEqual('[String!]!')
    expect(getNamedType(argType)).toEqual(GraphQLString)
  })

  it('Supports nested list argument type', () => {
    @InputObjectType()
    class Input {
      @InputField()
      bar: string
    }

    @ObjectType()
    class Foo {
      @Field()
      bar(@Arg({ type: [Input] }) input: Input[]): string {
        return 'ok'
      }
    }
    const { bar } = compileObjectType(Foo).getFields()
    const argType = bar.args[0].type
    expect(argType.toString()).toEqual('[Input!]!')
    expect(getNamedType(argType)).toEqual(compileInputObjectType(Input))
  })
})
