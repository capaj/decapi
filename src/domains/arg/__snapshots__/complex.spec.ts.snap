// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Complex arguments should not allow complex argument type decorated with @ObjectType 1`] = `"@Type Foo.bar(input <-------): Argument has incorrect type. Make sure to use native GraphQLInputType, native scalar like 'String' or class decorated with @InputObjectType"`;

exports[`Complex arguments should not allow complex argument type not decorated with @InputObjectType 1`] = `"@Type Foo.bar(input <-------): Could not infer type of argument. Make sure to use native GraphQLInputType, native scalar like 'String' or class decorated with @InputObjectType"`;
